client/cli/src/params/node_params.rs

```diff
@@ -0,0 +1,531 @@
+			self.rpc_methods,
+			self.validator,
+		)?;
+
+		Ok(Some(SocketAddr::new(
+			interface,
+			self.rpc_port.unwrap_or(9933),
+		)))
+	}
+
+	/// Get the RPC IPC path (`None` if disabled).
+	pub fn rpc_ipc(&self) -> Option<String> {
+		self.ipc_path.clone()
+	}
+
+	/// Get the RPC websocket address (`None` if disabled).
+	pub fn rpc_ws(&self) -> Result<Option<SocketAddr>> {
+		let interface = rpc_interface(
+			self.ws_external,
+			self.unsafe_ws_external,
+			self.rpc_methods,
+			self.validator,
+		)?;
+
+		Ok(Some(SocketAddr::new(
+			interface,
+			self.ws_port.unwrap_or(9944),
+		)))
+	}
+
+	/// Returns the RPC method set to expose.
+	pub fn rpc_methods(&self) -> sc_service::config::RpcMethods {
+		self.rpc_methods.into()
+	}
+
+	/// Get maximum runtime instances
+	pub fn max_runtime_instances(&self) -> Option<usize> {
+		self.max_runtime_instances.map(|x| x.min(256))
+	}
+
+	/// Get the base path of the configuration (if any)
+	pub fn base_path(&self) -> Result<Option<BasePath>> {
+		Ok(if self.tmp {
+			Some(BasePath::new_temp_dir()?)
+		} else {
+			None
+		})
+	}
+}
+
+fn rpc_interface(
+	is_external: bool,
+	is_unsafe_external: bool,
+	rpc_methods: RpcMethods,
+	is_validator: bool,
+) -> Result<IpAddr> {
+	if is_external && is_validator && rpc_methods != RpcMethods::Unsafe {
+		return Err(Error::Input(
+			"--rpc-external and --ws-external options shouldn't be \
+		used if the node is running as a validator. Use `--unsafe-rpc-external` \
+		or `--rpc-methods=unsafe` if you understand the risks. See the options \
+		description for more information."
+				.to_owned(),
+		));
+	}
+
+	if is_external || is_unsafe_external {
+		if rpc_methods == RpcMethods::Unsafe {
+			log::warn!(
+				"It isn't safe to expose RPC publicly without a proxy server that filters \
+			available set of RPC methods."
+			);
+		}
+
+		Ok(Ipv4Addr::UNSPECIFIED.into())
+	} else {
+		Ok(Ipv4Addr::LOCALHOST.into())
+	}
+}
+
+#[derive(Debug)]
+enum TelemetryParsingError {
+	MissingVerbosity,
+	VerbosityParsingError(std::num::ParseIntError),
+}
+
+impl std::error::Error for TelemetryParsingError {}
+
+impl std::fmt::Display for TelemetryParsingError {
+	fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+		match &*self {
+			TelemetryParsingError::MissingVerbosity => write!(f, "Verbosity level missing"),
+			TelemetryParsingError::VerbosityParsingError(e) => write!(f, "{}", e),
+		}
+	}
+}
+
+fn parse_telemetry_endpoints(s: &str) -> std::result::Result<(String, u8), TelemetryParsingError> {
+	let pos = s.find(' ');
+	match pos {
+		None => Err(TelemetryParsingError::MissingVerbosity),
+		Some(pos_) => {
+			let url = s[..pos_].to_string();
+			let verbosity = s[pos_ + 1..]
+				.parse()
+				.map_err(TelemetryParsingError::VerbosityParsingError)?;
+			Ok((url, verbosity))
+		}
+	}
+}
+
+/// CORS setting
+///
+/// The type is introduced to overcome `Option<Option<T>>`
+/// handling of `structopt`.
+#[derive(Clone, Debug)]
+pub enum Cors {
```

Comment:
Please move to its own file.

Reply:
Fair point, we have a file for those actually

Reply:
done in 1599a90